### PR TITLE
Advanced ICMP options

### PR DESCRIPTION
**Pull request by Gouthamve**
@RichiH wanted to add the payload size and DF-bit to be configurable in the ICMP prober. While I had fun implementing this, I would like to know if this is okay before I put in more effort.

cc/ @brian-brazil @SuperQ

---
This change is [Reviewable](https://reviewable.io/reviews/prometheus/blackbox_exporter/240)

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -96,7 +96,8 @@ type TCPProbe struct {
 
 type ICMPProbe struct {
 	PreferredIPProtocol string `yaml:"preferred_ip_protocol,omitempty"` // Defaults to "ip6".
-
+	Payload             int    `yaml:"payload,omitempty"`
```
Comment (brian-brazil) [inline config/config.go]: PayloadSize would make more sense

File: config/config.go [hunk 2]
```diff
@@ -96,7 +96,8 @@ type TCPProbe struct {
...
+	DontFragment        bool   `yaml:"dontfragment,omitempty"`
```
Comment (brian-brazil) [inline config/config.go]: dont_fragment

File: prober/icmp.go [hunk 3]
```diff
@@ -58,27 +54,27 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
...
 	defer socket.Close()
 
+	var data []byte
+	payload := module.ICMP.Payload
+	if payload != 0 {
+		data = make([]byte, payload)
+		copy(data, "Prometheus Blackbox Exporter")
+	} else {
+		data = []byte("Prometheus Blackbox Exporter")
+		payload = 1500 // The usual MTU size.
```
Comment (brian-brazil) [inline prober/icmp.go]: Where is this used?

File: prober/icmp.go [hunk 4]
```diff
@@ -136,3 +132,96 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
 		}
 	}
 }
+
+func getICMPSocket(
```
Comment (brian-brazil) [inline prober/icmp.go]: I don't think you're buying much by breaking this out, and it also makes it more difficult to review what has changed in the code

File: prober/icmp.go [hunk 5]
```diff
@@ -136,3 +132,96 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
...
+
+	return
+}
+
+type dfConn struct {
+	c *ipv4.RawConn
+}
+
+func (c *dfConn) ReadFrom(b []byte) (int, net.Addr, error) {
+	h, p, _, err := c.c.ReadFrom(b)
```
Comment (brian-brazil) [inline prober/icmp.go]: This won't work on Windows, which is a supported platform.

**Comment by SuperQ** [comment]
Feature idea seems good to me.

File: prober/icmp.go (hunk 5 above)
Comment (Gouthamve) [reply prober/icmp.go]: How do you propose I move about this? Different files for nix and windows? Or error when reading config?

**Comment by Gouthamve** [comment]
Updated to incorporate feedback and to error if `DontFragment` is set on windows. Don't have a windows machine to test though.

File: prober/icmp.go (hunk 5 above)
Comment (brian-brazil) [reply prober/icmp.go]: First see if this can be done with Windows, we should be aiming for feature parity

File: CONFIGURATION.md [hunk 6]
```diff
@@ -168,6 +168,12 @@ validate_additional_rrs:
 # The preferred IP protocol of the ICMP probe (ip4, ip6).
 [ preferred_ip_protocol: <string> | default = "ip6" ]
 
+# Set the DF-Bit in the IP-header. Only works with ip4 and on *nix systems.
```
Comment (brian-brazil) [inline CONFIGURATION.md]: bit

File: prober/icmp.go [hunk 7]
```diff
@@ -62,23 +62,44 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
...
+
 	defer socket.Close()
 
 	var data []byte
-	if payload != 0 {
-		data = make([]byte, payload)
-		copy(data[:], "Prometheus Blackbox Exporter")
+	if module.ICMP.PayloadSize != 0 {
+		data = make([]byte, module.ICMP.PayloadSize)
+		copy(data, "Prometheus Blackbox Exporter")
```
Comment (brian-brazil) [inline prober/icmp.go]: What if the payload size is smaller than this string?

Comment (Gouthamve) [reply prober/icmp.go]: https://play.golang.org/p/qrDFrlW6He Then only the payload size will be sen't with truncated string.

File: prober/icmp.go [hunk 8]
```diff
@@ -62,23 +62,52 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
...
+			socket = &dfConn{c: rc}
+		}
 	}
+
 	defer socket.Close()
 
+	var data []byte
+	if module.ICMP.PayloadSize != 0 {
+		data = make([]byte, module.ICMP.PayloadSize)
+		copy(data, "Prometheus Blackbox Exporter")
```
Comment (RichiH) [inline prober/icmp.go]: Nitpick, but it might make sense to include the exporter version in here.

**Pull request merged**